src/main/kotlin/org/wfanet/virtualpeople/core/selector/VidModelSelector.kt

```diff
@@ -0,0 +1,314 @@
+      require(modelLineId == modelRolloutModelLineId) {
+        "ModelRollouts must be parented by the provided ModelLine."
+      }
+    }
+  }
+
+  /**
+   * The adoption percentage of models is calculated each day. It consists of an array of triples
+   * where each triple wraps the lower and upper percentage bound for a particular ModelRelease. All
+   * the triples together cover the totality of events for a given day. The lower bound is
+   * inclusive. The upper bound is exclusive.
+   *
+   * E.g. [<0.0,0.5,ModelRelease1>,<0.5,1.1,ModelRelease2>] means that if the reducedEventId is
+   * lower than 0.5, ModelRelease1 is returned, ModelRelease2 otherwise.
+   */
+  private val lruCache: LruCache<Long, ArrayList<Triple<Double, Double, String>>> =
+    LruCache(CACHE_SIZE)
+
+  fun getModelRelease(labelerInput: LabelerInput): String? {
+    val eventTimestampSec = labelerInput.timestampUsec / 1_000_000L
+    if (
+      eventTimestampSec >= modelLine.activeStartTime.seconds &&
+        eventTimestampSec < modelLine.activeEndTime.seconds
+    ) {
+      val eventFingerprint = getEventFingerprint(labelerInput)
+      val reducedEventId = abs(eventFingerprint.toDouble() / Long.MAX_VALUE)
+      val eventDay: Long = eventTimestampSec / 86_400L
+      val ranges = readFromCache(eventDay)
+      for (triple in ranges) {
+        if (reducedEventId < triple.second) {
+          return triple.third
+        }
+      }
+    }
+    return null
+  }
+
+  /**
+   * Access to the cache is synchronized to prevent multiple threads calculating ranges in case of
+   * cache miss.
+   */
+  private fun readFromCache(eventDay: Long): ArrayList<Triple<Double, Double, String>> {
+    synchronized(this) {
+      if (lruCache.containsKey(eventDay)) {
+        return lruCache[eventDay]!!
+      } else {
+        val ranges = calculateRanges(eventDay)
+        lruCache[eventDay] = ranges
+        return ranges
+      }
+    }
+  }
+
+  /**
+   * A single triple with range 0.0..1.1 is returned if there is a single active ModelRollout. In
+   * case there are multiple active ModelRollout(s), the function iterates through an ordered list
+   * of active ModelRollout(s). The list is ordered by rollout_period_start_time, from the most
+   * recent to the oldest.
+   *
+   * The adoption percentage of each ModelRollout is calculated as follows: ((EVENT_DAY
+   * - ROLLOUT_START_DAY) / (ROLLOUT_END_DAY - ROLLOUT_START_DAY) - (1 *
+   *   percentage_taken_by_previous_rollout))
+   *
+   * In case a ModelRollout has the `rollout_freeze_time` set and the event day is greater than
+   * rollout_freeze_time, the EVENT_DAY in the above formula is replaced by `rollout_freeze_time` to
+   * ensure that the rollout stops its expansion. ((ROLLOUT_FREEZE_TIME - ROLLOUT_START_DAY) /
+   * (ROLLOUT_END_DAY - ROLLOUT_START_DAY) - (1 * percentage_taken_by_previous_rollout))
+   */
+  private fun calculateRanges(eventDay: Long): ArrayList<Triple<Double, Double, String>> {
```

Comment:
To me this code seems to be different than what we have converged on during the meetings.
The sequential roll-out that (I thought) we agreed upon would look like this pseudocode:
```
use_model = first known rollout
for each rollout ordered by start time:
  let percentage = (current time - rollout start) / (rollout end - rollout start)
  let fingerprint = Hash(Concatenate(rollout id, event fingerprint))
  if fingerprint < percentage:
    use_model = this rollout
```
While this implementation seems to be following a different algorithm.